### DISCUSSION TITLE
Music App Browse Requests

### DISCUSSION THREAD

**Discussion by hufman** [BimmerGestalt/AAIdrive#280] [Ideas]
Music apps can grant various levels of control over the Android MediaBrowser protocol. The music app can ❌ fully deny access, 
▶️ grant access to start the app but with an empty library, or ✔️ grant full control and browse access. If the music app grants access (with or without browsing) or if the user grants Notification Access, the app has the ability to control playback, get song metadata, and view any current queue. The available features for control are dependent on the music app, some provide cover art, some provide a queue, and so on.

If a music app ❌ denies access, there is nothing that can be done (without rooting the phone, cracking the music app, or implementing a custom API such as Spotify if available)

If the music app ▶️ provides access to start but with an empty library, sometimes it is possible to [hard-code an app-specific tweak](https://github.com/hufman/AndroidAutoIdrive/blob/e5968c276b587537e12f791444386fd82ff3dbbf/app/src/main/java/me/hufman/androidautoidrive/music/MusicBrowser.kt#L80) to gain access to the library. I occasionally look through the analytics to find candidate music apps, but I want you to suggest your favorite music apps that are Fully Controllable but not Browseable, and I'll investigate them and see if they can be enabled in this way.
![image](https://user-images.githubusercontent.com/1592375/118917137-0b355c00-b8e5-11eb-947e-58b4b51c1137.png)

- Manchester United - Only provides an empty library
- Soundcloud - Does not implement browse or queue at all
- WSJ - Only provides an empty library

**Comment by Kay0002** [comment]
Hi - browsing feature would be awesome for
Tidal
Pandora
IHeartRadio

**Reply by hufman** [reply]
I agree, I wish more apps would provide browsing by default! Unfortunately, apps that don't allow AAIdrive to connect at all aren't able to be helped:
![image](https://user-images.githubusercontent.com/1592375/136503037-de220423-265c-412c-828a-577f9cdd7553.png)

They close the connection before AAIdrive is allowed to browse the root folder, and so it's impossible to hardcode the proper root to browse the library.

**Reply by Kay0002** [reply]
Sorry, I didn't notice you responded sooner. I thought Tidal was fully controllable, I used to be able to browse new playlists and artists (etc), but I have not been able to browse anything as of late.

**Reply by hufman** [reply]
Unfortunately it seems that Tidal [closed access](https://github.com/BimmerGestalt/AAIdrive/issues/415) at some point. Their code theoretically opens up control to apps with Notification Access, but it didn't seem to provide browse access when I tried.

**Reply by Kay0002** [reply]
Well, I suppose it's true...All good things must come to an end. lol

I appreciate your quick responses as well as the fantastic job you've done with the AAIdrive app!

**Comment by blerglemon** [comment]
Hi! I think overall Plexamp works better than the Plex app itself, which causes a lot more issus than AAIdrive

**Reply by hufman** [reply]
Does PlexAmp work for you? It doesn't seem to work for me, it shows up as not available for AAIdrive to have full control, while Plex lets AAIdrive start it  :\

**Comment by hufman** [comment]
For my notes, the Sentry api can be queried to find a list of recently discovered music apps that match this criteria (connectable but not browseable):
```
curl -H "Authorization: Bearer e5a..." 'https://sentry.io/api/0/issues/1006330320/events/?&cursor=0:900:0' |
   jq -c '[.[] | {"tags":.tags| map({(.key): .value}) | add }] | .[] | select(.tags.connectable == "true") | select( .tags.browseable == "false") | {"appId":.tags.appId, "appName":.tags.appName}' |
   sort | uniq -c
```

**Comment by hufman** [comment]
QoBuz was found and fixed in #644
